### PR TITLE
Add 2 New Rules for Footnotes

### PR DESCRIPTION
**Pull request by chrisgrieser**
see issue #39. Also ran `npm run compile` and `npm run lint` this time :)

### CODE REVIEW

**Comment by chrisgrieser** [comment]
I would have added #36 as well, but I do not know how to get the `tabsize` setting.

**Comment by chrisgrieser** [comment]
added further exceptions for the capitalize rule

**Review by platers — CHANGES_REQUESTED** [review]

**Comment by platers** [comment]
> I would have added #36 as well, but I do not know how to get the `tabsize` setting.

You can pass it in as an option, the user can set `tabsize=2` in the settings. See the other rules for examples.

File: rules.ts [hunk 1]
```diff
@@ -55,7 +55,7 @@ export const rules: Rule[] = [
             'Removes trailing spaces and tabs',
             dedent`
         # H1   
-        line with trailing spaces and tabs    				`, // eslint-disable-line no-tabs
```
[the remark this answers is not in the document]
Comment (chrisgrieser) [reply rules.ts:58]: I actually didn't intend to? I think that could have happened by accident. (My keyboard is a bit broken, and I often do `undo` to fix stuff, my best guess I accidentally did something. -.-)

File: rules.ts [hunk 2]
```diff
@@ -153,7 +153,7 @@ export const rules: Rule[] = [
       `,
             dedent`
       # H1
-
```
[the remark this answers is not in the document]
Comment (chrisgrieser) [reply rules.ts:156]: also unintended

File: rules.ts [hunk 3]
```diff
@@ -97,13 +97,13 @@ export const rules: Rule[] = [
         `,
             dedent`
         # H1
-
```
[the remark this answers is not in the document]
Comment (chrisgrieser) [reply rules.ts:100]: also unintended

File: rules.ts [hunk 4]
```diff
@@ -508,6 +509,116 @@ export const rules: Rule[] = [
...
   ),
+  new Rule(
+      'Move Footnotes to the bottom',
+      'Move all footnotes to the bottom of the document.',
+      (text: string) => {
+        const footnotes = text.match(/^\[\^\w+\]: .*$/gm); // collect footnotes
+        if (footnotes != null) {
+          text = text.replace(/^\[\^\w+\]: .*$/gm, ''); // remove the footnotes
+          text += '\n\n' + footnotes.join('\n'); // append footnotes at the very end of the note
+          text = text.replace(/\n{3,}/gm, '\n\n'); // remove blank lines resulting from ft removal
```
[the remark this answers is not in the document]
Comment (chrisgrieser) [reply rules.ts:520]: see issue #42. 

you are right, it does also remove blank lines, de facto executing `consecutive-blank-lines` when executing rule. Should this be changed not to do that?

File: rules.ts [hunk 5]
```diff
@@ -508,6 +509,116 @@ export const rules: Rule[] = [
         ),
       ],
   ),
+  new Rule(
+      'Move Footnotes to the bottom',
+      'Move all footnotes to the bottom of the document.',
+      (text: string) => {
```
[the remark this answers is not in the document]
Comment (chrisgrieser) [reply rules.ts:515]: I tested the function outside of Obsidian with a text file and it didn't move anything when footnotes are already at the bottom.

How do I conduct such a test with test.ts? The readme states:
> Run `npm run compile` to build, generate documentation, and test the plugin.
I have run `npm run compile` before committing and it passed.

**Comment by chrisgrieser** [comment]
Okay just to be sure, I took the `rules.ts` from your repo, redid all the necessary steps again and commited once more. I have no idea what is/was happening with those tests that changed – I literally did not touch any of those things code sections that are supposedly changed. 

Anyway, this here seems to work, all tests pass when compiling, moving footnotes to the bottom does not do anything when already at the bottom. Depending on what you think about the issue with the consecutive blank lines when moving footnotes, I think this should work (other than the tests that are changed for some weird reason.)

File: rules.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (platers) [reply rules.ts:520]: I just dont want the rule not have unintended side effects, can it only remove lines where there used to be a footnote?

[the remark this answers is not in the document]
Comment (chrisgrieser) [reply rules.ts]: done

**Comment by chrisgrieser** [comment]
okay, modified the moving to bottom rule, now I think everything should be fine

**Review by platers — APPROVED** [review]

**Pull request merged**